### PR TITLE
Cr263 switch gcs datastore to firestore

### PR DESCRIPTION
**Pull request by PeterBochel**
# Motivation and Context
There are 2 main reasons to switch to Firestore from GCS (Google Cloud Storage):
1. Allow retrieval of cases by UPRN.
2. Performance.

# What has changed
The most significant change has been that GCSDataStore.java has been replaced with FirestoreDataStore.java. 

This switch to Firestore has caused relatively minor changes to ripple up to the CloudDataStore interface.

Test cases have been updated to test FirestoreDataStore. 

I've also added a component level integration test, Firestore_IT.java, which does testing of FirestoreDataStore with no mocking. So data is held by your Firestore datastore. These tests are currently marked as '@Ignore' as you really need to manually set things up to run this test.

# How to test?
1. Run all unit tests. 
2. Re-enable the tests in Firestore_IT and run that.
3. Manually run in UAC and CASE data. The steps to run a manual test have been added to the top level Readme.md file in the RH project.

# Links
See: https://collaborate2.ons.gov.uk/jira/browse/CR-263

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+      throws CTPException {
+    log.debug("Saving object to Firestore. Schema '{}' with key '{}'", schema, key);
+
+    // Store the object
+    ApiFuture<WriteResult> result = firestore.collection(schema).document(key).set(value);
+
+    // Wait for Firestore to complete
+    try {
+      result.get();
+      log.debug("Firstore save completed. Schema '{}' with key '{}'", schema, key);
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java]: typo

Comment (philwhiles) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java]: A Firstore is a seasonal shop, often a garage forecourt, selling a specific type of tree

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    // Run a query for a custom search path
+    FieldPath searchPath = FieldPath.of(fieldPath);
+    List<T> r = runSearch(target, schema, searchPath, searchValue);
+    log.debug("Firestore search return {} results", r.size());
+
+    return r;
+  }
+
+  private <T> List<T> runSearch(
+      Class<T> target, final String schema, FieldPath fieldPath, String searchValue)
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java]: suggest consistent naming fieldPath/searchPath

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceImpl.java [hunk 3]
```diff
@@ -30,167 +27,111 @@
...
+    String[] searchByUprnPath = new String[] {"address", "uprn"};
+    List<CollectionCase> searchResults =
+        cloudDataStore.search(CollectionCase.class, caseSchema, searchByUprnPath, uprn);
+
+    Optional<CollectionCase> collectionCase;
+    if (searchResults.isEmpty()) {
+      collectionCase = Optional.empty();
+    } else if (searchResults.size() == 1) {
+      collectionCase = Optional.of(searchResults.get(0));
+    } else {
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceImpl.java]: Will be a valid scenario to have more than one case for a upon (Household and Individuals at same address, flat 1a and flat flat 1b etc)

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+  /**
+   * Write object to Firestore collection. If the object already exists then it will be overwritten.
+   *
+   * @param schema - holds the name of the collection that the object will be added to.
+   * @param key - identifies the object within the collection.
+   * @param value - is the object to be written to Firestore. It must provide get methods that allow
+   *     access to its contents.
+   * @throws CTPException if any failure was detected interacting with Firestore.
+   */
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java]: java doc what type of object we can and cannot store

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+   * @param target is the object type that results should be returned in.
+   * @param schema is the schema to search.
+   * @param fieldPath is an array of strings that describe the path to the search field. eg, [
+   *     "case", "addresss", "postcode" ]
+   * @param searchValue is the value that the field must equal for it to be returned as a result.
+   * @return An optional which contains a List of results.
+   * @throws CTPException if anything goes wrong.
+   */
+  public <T> List<T> search(
+      Class<T> target, final String schema, String[] fieldPath, String searchValue)
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java]: "fieldPathElements"

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/RHSvcApplication.java [hunk 6]
```diff
@@ -69,4 +75,17 @@ public EventPublisher eventPublisher(final ConnectionFactory connectionFactory)
   public RestExceptionHandler restExceptionHandler() {
     return new RestExceptionHandler();
   }
+
+  /**
+   * Connect to Google Firestore.
+   *
+   * @return a Firestore connection.
+   */
+  @Bean
```
Comment (philwhiles) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/RHSvcApplication.java]: As discussed does this need to be outside/away from the FireStoreDataStore that actually uses it?

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java (hunk 1 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java]: Fixed.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java (hunk 2 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java]: Renamed as suggested.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java (hunk 4 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java]: Javadoc updated.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java (hunk 5 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/cloud/FirestoreDataStore.java]: Renamed as suggested.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceImpl.java (hunk 3 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/RespondentDataServiceImpl.java]: Method now returns a list of matching cases.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/RHSvcApplication.java (hunk 6 above)
Comment (PeterBochel) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/RHSvcApplication.java]: Have move the creation of the Firestore connection into the FirestoreDataStore class. Helps to contain the everything related to Firestore in just one class.

**Pull request merged**